### DISCUSSION TITLE
How to pass alpinejs variable to livewire component

### DISCUSSION THREAD

**Discussion by LittleHans8** [livewire/livewire#5390] [1. Help] [answered]
like this:
```
// questions is a alpinejs variable.
<template x-for="question in questions">
      <livewire:question-com :question_id="question.id">
</template>
```

**Comment by ousid** [comment]
Can you explain more, why you need to pass JS var to livewire component?

Not sure, but try this:

```blade
<template x-for="question in questions">
    <livewire:question-com question_id="@{{ question.id }}"/>
</template>
```

You may use `@entangle` magic property, here is a link to know [more](https://laravel-livewire.com/docs/2.x/alpine-js#sharing-state)

**Reply by LittleHans8** [reply]
Since I have a lot of repetitive components that need to be rendered into html, rendering html on the front-end page with alpinejs is much more efficient than returning the entire html on the server side with livewire, and can reduce the size by more than half.

**Comment by joshhanley** [comment, marked as answer]
@LittleHans8 the reason this won't work, is because `<livewire:... />` tags are rendered on the server by blade. But you are dynamically adding them to the page using javascript, which is happening after the blade render has finished.

The only way generate Livewire components in a loop is to use a blade `@foreach` statement.

Personally I wouldn't worry too much about the size, as that html will only be loaded on the initial render. For each subsequent render of the parent component, Livewire doesn't re-render the child components, so only the html of the parent will be sent, with empty html where the child component would be. And each child component will only re-render if a request is sent to it specifically, and none of the others will. 

But for that to work, you need to make sure you use `wire:key` on your child components. See https://laravel-livewire.com/docs/2.x/nesting-components#sibling-components-in-a-loop

Hope this helps!

**Reply by LittleHans8** [reply]
Thank you for your answer, that's what I'm doing on my current production site.

But some users feedback that the site is slow to load, and I have located that the problem is that I need to load nearly 1000 duplicate components at one time, and I can't paginate them for user needs, plus the css style of the components, the size of the html page returned at one time is too large, close to 1MB.

So my idea is to render these duplicate components in the front-end page, then only 1 component in the returned html needs to be transferred, the remaining 999 duplicate components can be rendered in the front-end page, and the total size of the html will be reduced to about 300kb.

From your answer, it looks like the back-end components are not rendered on the front-end page, so maybe I need to optimize from somewhere else

**Reply by ousid** [reply]
If the logic does not require a heavy interaction with the backend, I think it's better to just create a basic Laravel component, and use it.

Not every case requires a livewire component (Personal opinion).

**Reply by LittleHans8** [reply]
> If the logic does not require a heavy interaction with the backend, I think it's better to just create a basic Laravel component, and use it.
> 
> Not every case requires a livewire component (Personal opinion).

Yes, that's exactly what I was thinking.

My current tentative optimization solution is to remove all components, render the duplicate content on the front-end page, switch to an event-listening [$wire.emit](https://laravel-livewire.com/docs/2.x/alpine-js#interacting-with-livewire-from-alpine) method for front and back-end interaction, and finally listen for content changes in JavaScript [Livewire.on](https://laravel-livewire.com/docs/2.x/events#in-js) to manually update parts of the UI.
 
I hope this solution is helpful for people who have the same problem, and if anyone has a better way, feel free to share it here.

**Comment by arapchirchir** [comment]
// questions is a alpinejs variable.

```
<template x-for="question in questions">
    //  <livewire:question-com :question_id="question.id">
    <div x-data="{ question:question}">
        @include('components.question-component')
     </div>
</template>
```

 This is what you can do
 
 and then in file question-component
```
 <template x-if="question">
    <div x-text="question.id"></div> 
</template>
```
